add gambit documentation (#54)
Certora/Documentation: docs/gambit/mutation-verifier.md

`--send_only` and `--cloud` in the run scripts.
If you have these flags, please remove them for now.**
Apologies for the temporary inconvenience!

The tool expects a configuration file (extension `.conf` is required)
which is a JSON object.
Here is an example configuration file:

```
{
    "project_folder" : "Test/10Power",
    "run_script" : "Test/10Power/runDefault.sh",
    "gambit" : {
        "filename" : "Test/10Power/TenPower.sol",
        "solc" : "solc8.10"
    },
    "staging" : "master",
    "manual_mutations" : {
        "path/to/original/file.sol" : "path/to/manual/mutations/",
        ...
    },
    ...
}
```

Required Keys for the JSON Configuration File:
- `"project_directory"` : the directory containing the original CVT project on which to perform mutation testing
- `"run_script"` : the bash script used to run verification on the original project, usually `project_directory/run.sh` or similar
- `"gambit"` : the JSON configuration element for invoking Gambit. May be a path to a gambit configuration file
or the explicit JSON element contained therein. Refer to the open source Gambit repository for more details regarding Gambit configuration files.

> Optional Keys for the JSON Configuration File:

Reviewer: move the internal Certora ones to the end, and maybe mark them as "for debugging"
--staging
--"use_cli_certora_run"`

Author: Done. A quick note about this though is that `--staging` is not default. By default it runs locally (because if you don't have CERTORAKEY, that can be a problem). Just FYI. We can change it if you think it's the right default but I would think it should be set by user.

Reviewer 2: For any user other than a Certora dev, we shouldn't expect them to have a local installation.  The documentation and standard usage should assume that users are running on the cloud with a `CERTORAKEY`, and we should only mention running locally in internal documentation (on confluence).

- `"num_threads"` : the maximum number of threads to use for verification, as an integer
- `"manual_mutations"` : optionally supplement the random mutant generation with your own manually-written mutants.
Expects a JSON object whose keys are the paths to the original files and whose values are paths to directories containing
manually-written mutants as `.sol` files. **IMPORTANT:** any manual mutations files provided must follow the naming
convention `OriginalFileName.<unique-name>.sol`, where `unique-name` is a string ID unique with respect to the other
manual mutants. It is recommended to use `mN` for brevity, where `N` is a unique integer.
- `"offline"` : run mutation testing without internet connection, skipping the UI output and other web functions.
Expects a boolean and defaults to `false`.

Additional Optional Flags for Debugging
- `"staging"` : the branch name for running mutant verification on `--staging`. We support `"staging" : true` as an
alternative to `"staging" : "master"`. Omitting this key will cause verification to run locally.
- `"use_cli_certora_run"` : Use CLI `certoraRun` rather than `certoraRun.py`. Expects a boolean and defaults to `false`.

> For implementation details regarding the generation of mutants, refer to the open source repository for Gambit.

Reviewer 3: Link to Gambit repo?

## Visualization

The mutation verification results are
  summarized in an user-friendly visualization.
[Here](https://mutation-testing-beta.certora.com/reports/mutation?id=c7c659d7-d500-46f2-acf1-1392eee714b5&anonymousKey=f4b40ba6-2160-4993-9f50-02625b291cae) is an example summary

>   for the [Borda example](https://github.com/Certora/CodeExamples/tree/master/Borda).

Author: so actually before I tried this repo, I had the link to `https://demo.certora.com/?Borda` and that also didn't work..

> # Solidity Mutation Verifier Based on Gambit

Author: yes. I think the top level description in index.md helps clarify this.

This is a mutation verifier which
 checks that variants of the original
 solidity program do not pass the specification.
If a mutated program passes the specification,
it may indicate that the specification is vacuous or not rigorous enough.
It uses mutations from the [Gambit](https://github.com/Certora/gambit)
  mutation generator.
In the rest of the document,
  we refer to both the mutation generator and the verifier as Gambit.

## Running the Mutation Verifier

- Example:
```

> java -ea -jar $CERTORA/certora_jars/MutationTest.jar /path/to/config/file/Example.conf

Author: Wherever the jars (like the type checker) get put for customers. For devs it's in the directory defined using `$CERTORA`. What is it for customers?

> java -ea -jar $CERTORA/certora_jars/MutationTest.jar /path/to/config/file/Example.conf

Author: yeah perhaps. Like i said, i think in a future PR, we can clarify this when we get someone else to install it. The idea is that when they get the certora-cli installed, they will get these binaries put in the "right" places.

```

- Remember to `yarn install` or `npm install` when working
  on customer code if required by the customer code infra!

- **NOTE: if a path has spaces, put quotes around it to ensure correct parsing by Kotlin's libraries**

> - Gambit **does** support `--staging`!

Reviewer 2: Link to the command-line options in the Prover manual using, e.g.
```
{ref}`--staging`
```

You may need to add a ref to `/docs/prover/cli/options.md`, search for `--rule_sanity` in that document for an example

Author: hmm if you can help with this, that would be great. Not sure I am following what you are saying.

**However, Gambit currently has trouble with
`--send_only` and `--cloud` in the run scripts.
If you have these flags, please remove them for now.**
Apologies for the temporary inconvenience!

The tool expects a configuration file (extension `.conf` is required)
which is a JSON object.
Here is an example configuration file:

```
{
    "project_folder" : "Test/10Power",
    "run_script" : "Test/10Power/runDefault.sh",
    "gambit" : {
        "filename" : "Test/10Power/TenPower.sol",
        "solc" : "solc8.10"
    },
    "staging" : "master",
    "manual_mutations" : {
        "path/to/original/file.sol" : "path/to/manual/mutations/",
        ...
    },
    ...
}
```

Required Keys for the JSON Configuration File:
- `"project_directory"` : the directory containing the original CVT project on which to perform mutation testing

> - `"run_script"` : the bash script used to run verification on the original project, usually `project_directory/run.sh` or similar

Reviewer 2: Does this work if the project is using certora conf files?  E.g. [the ERC20 Example](https://github.com/Certora/ERC20Example/blob/main/certora/scripts/verifyERC20.sh)

Author: no, it does not.

alternative to `"staging" : "master"`. Omitting this key will cause verification to run locally.
- `"use_cli_certora_run"` : Use CLI `certoraRun` rather than `certoraRun.py`. Expects a boolean and defaults to `false`.

For implementation details regarding the generation of mutants,
  refer to the open source repository
  for [Gambit](https://github.com/Certora/gambit).

## Visualization

The mutation verification results are
  summarized in an user-friendly visualization.
[Here](https://mutation-testing-beta.certora.com/reports/mutation?id=c7c659d7-d500-46f2-acf1-1392eee714b5&anonymousKey=f4b40ba6-2160-4993-9f50-02625b291cae) is an example summary
  for the [Borda example](https://github.com/Certora/CodeExamples/tree/master/Borda).
Rules are represented by the green outer circles
  and the mutants are represented by the gray dots.
Selecting a rule shows which mutants it killed
  and selecting a mutant shows which rules killed it.
The coverage metric is computed as the fraction
  of total generated mutants that were killed.
Clicking on a mutant's patch also shows the
  diff with respect to the original program.

> ## Implementation Details

Reviewer 2: If this is not relevant to users, it should probably not be included in the public docs

> # Using Gambit with the Prover

Reviewer 2: It would be really good to set up https://github.com/Certora/ERC20Example with a gambit config and use it for examples.

Author: yes! Let me do that as a separate PR if that is ok. Like I was saying, we are waiting for the packaging to be done after which this will be much easier.

This is the mutation verifier which
 checks that variants of the original
 solidity program do not pass the specification.
It uses mutations from the [Gambit](https://github.com/Certora/gambit)
  mutation generator.
It also allows users to include manually generated mutants.
If a mutated program passes the specification,
it may indicate that the specification is vacuous or not rigorous enough.
In the rest of the document,
  we refer to both the mutation generator and the verifier as Gambit.

## Installations and Setup

You will require java to run the mutation testing jar.
Installation instructions can be found [here](https://www.java.com/en/download/help/download_options.html).

> 

Author: I am not totally sure. But whenever they get the typechecking jar.

## Running the Mutation Verifier

- Example:
```
java -ea -jar $CERTORA/certora_jars/MutationTest.jar /path/to/config/file/Example.conf
```

> 

Reviewer 2: Something important that's missing here is the fact that gambit pulls the configuration for `certoraRun` from a shell script.  This deserves at least a paragraph.

Author: ok!

- **NOTE: if a path has spaces, put quotes around it to ensure correct parsing by Kotlin's libraries**

> - Gambit supports `--staging`!
> 
> **However, Gambit currently has trouble with
> `--send_only` and `--cloud` in the run scripts.
> If you have these flags, please remove them for now.**
> Apologies for the temporary inconvenience!

Reviewer 2: ````suggestion
```{info}
Gambit supports {ref}`--staging`.

However, Gambit currently has trouble with
{ref}`--send_only` and {ref}`--cloud` in the run scripts.
If you have these flags, please remove them for now.
Apologies for the temporary inconvenience!
```
````

> Apologies for the temporary inconvenience!

Author: so this is actually important. This means that the SE/SRs should not have these flags in their run scripts for now. I think this info should be visible as early as possible because in my experience, users have these flags most of the time.

> The tool expects a configuration file (extension `.conf` is required)

Reviewer 2: ```suggestion
(gambit-prover-config)=
## Configuration
The tool expects a configuration file (extension `.conf` is required).
```

> The tool expects a configuration file (extension `.conf` is required)

Author: Ah yeah. so thing is, Certora users don't have to write the gambit conf files. They can embed it in this conf file that we are describing here.

> The tool expects a configuration file (extension `.conf` is required)

Author: i will clarify this anyway.

which is a JSON object.
Here is an example configuration file:

```
{
    "project_folder" : "Test/10Power",
    "run_script" : "Test/10Power/runDefault.sh",
    "gambit" : {
        "filename" : "Test/10Power/TenPower.sol",
        "solc" : "solc8.10"
    },
    "staging" : "master",
    "manual_mutations" : {
        "path/to/original/file.sol" : "path/to/manual/mutations/",
        ...
    },
    ...
}
```

> Required Keys for the JSON Configuration File:

Reviewer 2: ```suggestion
### Required Keys for the JSON Configuration File:
```

- `"project_directory"` : the directory containing the original CVT project on which to perform mutation testing
- `"run_script"` : the bash script used to run verification on the original project, usually `project_directory/run.sh` or similar

> - `"gambit"` : the JSON configuration element for invoking Gambit. May be a path to a gambit configuration file

Reviewer 2: May be worth noting that some of the `solc` arguments will need to be here even if they are already in the run script (assuming that's true?)

Author: yup! agreed.

> or the explicit JSON element contained therein. Refer to the open source Gambit repository for more details regarding Gambit configuration files.

Reviewer 2: ```suggestion
or the explicit JSON element contained therein.  See {ref}`gambit-config` for more information about the gambit configuration.
```

Reviewer 2: You'll have to add `(gambit-config)=` on the line before the corresponding section in the other document for the `ref` to work.

> Optional Keys for the JSON Configuration File:

Reviewer 2: ```suggestion
### Optional Keys for the JSON Configuration File
```

- `"num_threads"` : the maximum number of threads to use for verification, as an integer
- `"manual_mutations"` : optionally supplement the random mutant generation with your own manually-written mutants.
Expects a JSON object whose keys are the paths to the original files and whose values are paths to directories containing

> manually-written mutants as `.sol` files. **IMPORTANT:** any manual mutations files provided must follow the naming

Reviewer 2: I would call this out with an
````
```{info}
info block
```
````

> convention `OriginalFileName.<unique-name>.sol`, where `unique-name` is a string ID unique with respect to the other

Reviewer 2: ```suggestion
convention `OriginalFileName.<unique-name>.sol`, where `<unique-name>` is a string ID unique with respect to the other
```

> manual mutants. It is recommended to use `mN` for brevity, where `N` is a unique integer.

Reviewer 2: ```suggestion
manual mutants (for example you might name them `OriginalFileName.m1.sol`, `OriginalFileName.m2.sol` and so on).
```

> - `"offline"` : run mutation testing without internet connection, skipping the UI output and other web functions.

Reviewer 2: This should be included in the debugging section, since non-internal users can't run locally.

Expects a boolean and defaults to `false`.

> Additional Optional Flags for Debugging

Reviewer 2: ```suggestion
### Additional Optional Flags for Certora Internal Use
```

> - `"staging"` : if your run script does not already have `--staging`, you can also add it to Gambit.

Reviewer 2: ```suggestion
- `"staging"` : if your run script does not already have {ref}`--staging`, you can also add it to Gambit.
```

  Similar to CVT, you can provide the
  branch name for running mutant verification on `--staging`.
We support `"staging" : true` as an alternative to `"staging" : "master"`.
Omitting this key will cause verification to run locally
  (unless the run script has it).
- `"use_cli_certora_run"` : Use CLI `certoraRun` rather than `certoraRun.py`. Expects a boolean and defaults to `false`.

> For implementation details regarding the generation of mutants,
>   refer to the open source repository
>   for [Gambit](https://github.com/Certora/gambit).

Reviewer 2: ```suggestion
```

I think this is unnecessary because we already mention this in `gambit.md` and we already link to that.

## Visualization

The mutation verification results are
  summarized in an user-friendly visualization.
[Here](https://mutation-testing-beta.certora.com/reports/mutation?id=c7c659d7-d500-46f2-acf1-1392eee714b5&anonymousKey=f4b40ba6-2160-4993-9f50-02625b291cae) is an example summary

>   for the [Borda example](https://demo.certora.com/?Borda).

Reviewer 2: This takes some time to load for me.  It's probably worth including a screenshot here.

Author: very strange.

Rules are represented by the green outer circles
  and the mutants are represented by the gray dots.

> Selecting a rule shows which mutants it killed

Reviewer 2: ```suggestion
Selecting a rule shows which mutants it detected
```

>   and selecting a mutant shows which rules killed it.

Reviewer 2: ```suggestion
  and selecting a mutant shows which rules detected it.
```

The coverage metric is computed as the fraction

>   of total generated mutants that were killed.

Reviewer 2: ```suggestion
  of total generated mutants that were detected.
```
